src/coreComponents/physicsSolvers/wavePropagation/AcousticWaveEquationSEM.cpp

```diff
@@ -0,0 +1,759 @@
+                                         receiverCoordinates[ircv][2] };
+
+              real64 coordsOnRefElem[3]{};
+              bool const receiverFound = computeCoordinatesOnReferenceElement< FE_TYPE >( coords, coordsOnRefElem, k, faceNodes, elemsToNodes, X );
+              if( receiverFound )
+              {
+                receiverIsLocal[ircv] = 1;
+
+                real64 Ntest[8];
+                finiteElement::LagrangeBasis1::TensorProduct3D::value( coordsOnRefElem, Ntest );
+
+                for( localIndex a=0; a< numNodesPerElem; ++a )
+                {
+                  receiverNodeIds[ircv][a] = elemsToNodes[k][a];
+                  receiverConstants[ircv][a] = Ntest[a];
+                }
+              }
+            }
+          } // End loop over receiver
+
+        } // End loop over elements
+      } );
+    } );
+  } );
+}
+
+
+void AcousticWaveEquationSEM::addSourceToRightHandSide( real64 const & time_n, arrayView1d< real64 > const rhs )
+{
+  arrayView2d< localIndex const > const sourceNodeIds = m_sourceNodeIds.toViewConst();
+  arrayView2d< real64 const > const sourceConstants   = m_sourceConstants.toViewConst();
+  arrayView1d< localIndex const > const sourceIsLocal = m_sourceIsLocal.toViewConst();
+  real64 fi = 0;
+  switch( this->m_rickerOrder )
+  {
+    case 2:
+    {
+      fi = evaluateRickerOrder2( time_n, this->m_timeSourceFrequency );
+    }
+    break;
+    case 1:
+    {
+      fi = evaluateRickerOrder1( time_n, this->m_timeSourceFrequency );
+    }
+    break;
+    case 0:
+    {
+      fi = evaluateRicker( time_n, this->m_timeSourceFrequency );
+    }
+    break;
+    default:
+      GEOSX_ERROR( "This option is not supported yet, rickerOrder must be 0, 1 or 2" );
+  }
+
+  for( localIndex isrc = 0; isrc < sourceConstants.size( 0 ); ++isrc )
+  {
+    if( sourceIsLocal[isrc] == 1 )
+    {
+      for( localIndex inode = 0; inode < sourceConstants.size( 1 ); ++inode )
+      {
+        rhs[sourceNodeIds[isrc][inode]] = sourceConstants[isrc][inode] * fi;
+      }
+    }
+  }
+}
+
+
+void AcousticWaveEquationSEM::computeSismoTrace( localIndex const isismo, arrayView1d< real64 > const pressure_np1 )
+{
+  arrayView2d< localIndex const > const receiverNodeIds = m_receiverNodeIds.toViewConst();
+  arrayView2d< real64 const > const receiverConstants   = m_receiverConstants.toViewConst();
+  arrayView1d< localIndex const > const receiverIsLocal = m_receiverIsLocal.toViewConst();
+
+  arrayView1d< real64 > const p_rcvs   = m_pressureNp1AtReceivers.toView();
+
+  for( localIndex ircv = 0; ircv < receiverConstants.size( 0 ); ++ircv )
+  {
+    if( receiverIsLocal[ircv] == 1 )
+    {
+      p_rcvs[ircv] = 0.0;
+      for( localIndex inode = 0; inode < receiverConstants.size( 1 ); ++inode )
+      {
+        p_rcvs[ircv] += pressure_np1[receiverNodeIds[ircv][inode]]*receiverConstants[ircv][inode];
+      }
+
+      if( this->m_outputSismoTrace == 1 )
```

Comment:
I would make this into a separate loop. That way you can make the above something other than `serialPolicy`.

Reply:
I split it into two separate forAll< serialPolicy > loop. The number of receivers should not be that much and the saveSismo is a temporary solution before we get the same functionality with timeHistory.